rfcs/0120-artifact-service.md

```diff
@@ -0,0 +1,296 @@
+
+The completion endpoint is used to perform any post-upload validation required.
+This might include the "Complete Multipart Upload" for S3 based objects, or
+possible a post-upload hash and verification step for a locally managed file
+server.
+
+Until the completion endpoint is called for a given object, attempts to
+retrieve an object will result in a 404 Not Found response.
+
+Example of a multipart upload using S3 from the object service's perspective.
+This is a simplified view, meant to highlight the key interactions rather than
+every detail.  For the purpose of this example, the uploader speaks directly
+with the object service.  In reality this interaction would have something like
+the Queue as an intermediary between the uploader and the object service.
+
+```
+uploader --> objsvc PUT /objects/abcd123  {
+                                            origin: '10.10.10.10',
+                                            contentType: 'application/json',
+                                            contentLength: 1234,
+                                            contentSha256: 'abcdef1234',
+                                            transferLength: 1234,
+                                            transferSha256: 'abcdef1234',
+                                            contentEncdoing: 
+                                            expiration: new Date(),
+                                            parts: [
+                                              sha256: 'abcd1234',
+                                              size: 1024,
+                                            }, {
+                                              sha256: 'abcd1234',
+                                              size: 1024,
+                                            }],
+                                          }
+objsvc --> s3 POST /object-abcd123?uploads
+objsvc <-- s3 200 UploadId=u-123
+uploader <-- objsvc 200 {requests: [{
+                          url: 'http://s3.com/object-abcd123?partNumber=1&uploadId=u-123',
+                          headers: {authorization: 'signature-request-1'},
+                          method: 'PUT'
+                        }, {
+                          url: 'http://s3.com/object-abcd123?partNumber=2&uploadId=u-123',
+                          headers: {authorization: 'signature-request-1'},
+                          method: 'PUT'
+                        }
+                       ]}
+uploader --> s3 PUT /object-abcd123?partNumber=1&uploadId=u-123'
+uploader <-- s3 200 ETag {etag: 123}
+uploader --> s3 PUT /object-abcd123?partNumber=2&uploadId=u-123'
+uploader <-- s3 200 ETag {etag: 456}
+uploader --> objsvc PATCH /objects/abcd123 {etags: [123, 456]}
+objsvc --> s3 POST /object-abcd123?uploadId=u-123 {etags: [123, 456]}
+objsvc <-- s3 200 ETag {etag: 123456}
+uploader <-- objsvc 200
+```
+
+### Deletetion
+
+```
+DELETE /objects/:name
+DELETE /caches/objects/:name
```

Comment:
If we need to purge caches, then isn't this an internal bug? I propose not supporting it :)

Or maybe change it to `DELETE /objects/:name?cacheOnly=true` so it at-least feels more REST-like.

Reply:
The latter seems good to me.  I think we've learned that building tools to support fixing things that shouldn't break is a good strategy :)

Reply:
Purging caches might need to be done for reasons outside of the object service's control.  Like deleting, let's build a full operational toolbox for those supporting the service by building a very likely to be useful tool with full unit tests.  Adding functionality like this is how we build services that can be operationally supported by someone other than the primary developers.

The problem with `cacheOnly=true` is that likely someone using something like curl is going to have command history for the get:

```
curl http://objects.tc.net/objects/my-object
```

then go up to that and add a `-X DELETE` and delete the real thing.  I say that because this is exactly what I would do by mistake, I guarantee it ;)

Since deletion of the canonical copy is likely very significantly more rarely desired than deleting a cached copy, if we wanted to use the same resource name, I would rather that it be a cache without the arguments, and require an option for the canonical resource.

I'm also not sure how this is less REST to have `DELETE /caches/objects/:name`.  There's nothing stopping us from having a `GET /caches` endpoint which gives a list of configured caches, or `GET /caches/status/:cacheId` which shows detail of a particular cache, or `GET /caches/objects/:name/status` which lists what the status of the object is in each of the caches.

Is there a specific technical reason for not using `/caches/objects`, or is this more of a preference thing?